File: lib/scanner.js
const fs = require('fs');
const path = require('path');

function scanSkills(sourceDir) {
  if (!fs.existsSync(sourceDir)) {
    return [];
  }

  const items = fs.readdirSync(sourceDir);
  const skills = [];

  items.forEach(item => {
    // 跳过隐藏文件和 .gitkeep
    if (item.startsWith('.') || item === '.gitkeep') {
      return;
    }

    const fullPath = path.join(sourceDir, item);
    const stat = fs.statSync(fullPath);

    // 添加文件或目录
    if (stat.isFile() || stat.isDirectory()) {
      skills.push({
        name: item,
        path: fullPath,
        isDirectory: stat.isDirectory()
      });
    }
  });

  return skills;
}

function scanCommands(sourceDir) {
  if (!fs.existsSync(sourceDir)) {
    return [];
  }

  const items = fs.readdirSync(sourceDir);
  const commands = [];

  items.forEach(item => {
    if (item.startsWith('.')) return;

    const fullPath = path.join(sourceDir, item);
    const stat = fs.statSync(fullPath);

    if (stat.isFile()) {
      if (!item.endsWith('.md')) return;
      commands.push({
        name: item,
        path: fullPath,
        isDirectory: false,
        parent: null
      });
    } else if (stat.isDirectory()) {
      const children = [];
      const subItems = fs.readdirSync(fullPath);

      subItems.forEach(subItem => {
        if (subItem.startsWith('.')) return;

        const subPath = path.join(fullPath, subItem);
        const subStat = fs.statSync(subPath);

        if (subStat.isDirectory()) {
          console.log(`⚠ 跳过二级嵌套: ${item}/${subItem}`);
          return;
        }

        if (subStat.isFile() && subItem.endsWith('.md')) {
          children.push(subItem);
        }
      });

      commands.push({
        name: item,
        path: fullPath,
        isDirectory: true,
        children
      });
    }
  });

  return commands;
}

function getSkillType(skillName) {
  return skillName.endsWith('/') ? 'dir' : 'file';
}

function flattenCommandName(folderName, fileName) {
  return `${folderName}-${fileName}`;
}

function expandCommandsForTool(commands, tool, subfolderSupport) {
  const expanded = [];

  commands.forEach(cmd => {
    if (!cmd.isDirectory) {
      expanded.push({
        name: cmd.name,
        sourcePath: cmd.path,
        isFlattened: false
      });
    } else if (subfolderSupport) {
      expanded.push({
        name: cmd.name,
        sourcePath: cmd.path,
        isDirectory: true,
        isFlattened: false
      });
    } else {
      cmd.children.forEach(child => {
        expanded.push({
          name: flattenCommandName(cmd.name, child),
          sourcePath: require('path').join(cmd.path, child),
          isFlattened: true,
          originalFolder: cmd.name
        });
      });
    }
  });

  return expanded;
}

module.exports = {
  scanSkills,
  scanCommands,
  getSkillType,
  flattenCommandName,
  expandCommandsForTool
};

File: python_app/core/scanner.py
from pathlib import Path

from .markdown_description import read_description_from_markdown, read_description_from_skill_folder

def scan_skills(source_dir: str) -> list[dict[str, object]]:
    source_path = Path(source_dir)
    if not source_path.exists():
        return []
    skills: list[dict[str, object]] = []
    for item in sorted(source_path.iterdir(), key=lambda entry: entry.name.lower()):
        if item.name.startswith(".") or item.name == ".gitkeep":
            continue
        if item.is_file() or item.is_dir():
            meta = read_description_from_skill_folder(item)
            skills.append(
                {
                    "name": item.name,
                    "path": str(item),
                    "isDirectory": item.is_dir(),
                    "description": meta.description if meta else "",
                    "descriptionSource": meta.source if meta else "",
                }
            )
    return skills


def scan_commands(source_dir: str) -> list[dict[str, object]]:
    source_path = Path(source_dir)
    if not source_path.exists():
        return []
    commands: list[dict[str, object]] = []
    for item in sorted(source_path.iterdir(), key=lambda entry: entry.name.lower()):
        if item.name.startswith("."):
            continue
        if item.is_file() and item.suffix == ".md":
            meta = read_description_from_markdown(item)
            commands.append(
                {
                    "name": item.name,
                    "path": str(item),
                    "isDirectory": False,
                    "parent": None,
                    "children": [],
                    "description": meta.description if meta else "",
                    "descriptionSource": meta.source if meta else "",
                }
            )
            continue
        if not item.is_dir():
            continue
        children = [
            child.name
            for child in sorted(item.iterdir(), key=lambda entry: entry.name.lower())
            if child.is_file() and child.suffix == ".md" and not child.name.startswith(".")
        ]
        summary = f"包含 {len(children)} 个子命令" if children else ""
        commands.append(
            {
                "name": item.name,
                "path": str(item),
                "isDirectory": True,
                "parent": None,
                "children": children,
                "description": summary,
                "descriptionSource": "derived" if summary else "",
            }
        )
    return commands


def flatten_command_name(folder_name: str, file_name: str) -> str:
    return f"{folder_name}-{file_name}"


def expand_commands_for_tool(
    commands: list[dict[str, object]],
    tool_id: str,
    subfolder_support: bool,
) -> list[dict[str, object]]:
    del tool_id
    expanded: list[dict[str, object]] = []
    for command in commands:
        if not command["isDirectory"]:
            expanded.append(
                {
                    "name": command["name"],
                    "sourcePath": command["path"],
                    "isFlattened": False,
                }
            )
            continue
        if subfolder_support:
            expanded.append(
                {
                    "name": command["name"],
                    "sourcePath": command["path"],
                    "isDirectory": True,
                    "isFlattened": False,
                }
            )
            continue
        for child in command["children"]:
            expanded.append(
                {
                    "name": flatten_command_name(command["name"], child),
                    "sourcePath": str(Path(command["path"]) / child),
                    "isFlattened": True,
                    "originalFolder": command["name"],
                }
            )
    return expanded
